lib/calculate_reference_proteome_similarity.py

```diff
@@ -0,0 +1,95 @@
+        flanking_sequence_length = 7
+        mt_start = (subpeptide_position-1) + (mutation_position-1)
+        start = mt_start - flanking_sequence_length
+        if start < 0:
+            start = 0
+        end = mt_start + mt_length + flanking_sequence_length
+        return full_peptide[start:end]
+
+    def extract_n_mer_from_fs(self, full_peptide, epitope, peptide_sequence_length, subpeptide_position):
+        if peptide_sequence_length%2 == 0:
+            flanking_sequence_length = int(peptide_sequence_length/2)
+        else:
+            flanking_sequence_length = int((peptide_sequence_length-1)/2)
+        start = subpeptide_position - 1 - flanking_sequence_length
+        if start < 0:
+            start = 0
+        end = subpeptide_position + len(epitope) + flanking_sequence_length
+        return full_peptide[start:end]
+
+    def execute(self):
+        if self.species not in self.species_to_organism:
+            print("Species {} not supported for Reference Proteome Similarity search. Skipping.".format(self.species))
+            shutil.copy(self.input_file, self.output_file)
+            return
+
+        records_dict = self.get_peptides()
+
+        with open(self.input_file) as input_fh, open(self.output_file, 'w') as output_fh:
+            reader = csv.DictReader(input_fh, delimiter = "\t")
+            writer = csv.DictWriter(output_fh, delimiter = "\t", fieldnames=reader.fieldnames + self.reference_match_headers(), extrasaction='ignore')
+            writer.writeheader()
+            for line in reader:
+                if self.file_type == 'pVACbind':
+                    epitope = line['Epitope Seq']
+                else:
+                    epitope = line['MT Epitope Seq']
+                if line['Variant Type'] == 'FS':
+                    peptide = self.extract_n_mer_from_fs(records_dict[line['Index']], epitope, self.peptide_sequence_length, int(line['Sub-peptide Position']))
+                else:
+                    mt_amino_acids = line['Mutation'].split('/')[1]
+                    if mt_amino_acids == '-':
+                        mt_amino_acids = ''
+                    peptide = self.extract_n_mer(records_dict[line['Index']], int(line['Sub-peptide Position']), int(line['Mutation Position']), len(mt_amino_acids))
+                result_handle = NCBIWWW.qblast("blastp", "refseq_protein", peptide, entrez_query="{} [Organism]".format(self.species_to_organism[self.species]))
+                reference_match = False
+                for blast_record in NCBIXML.parse(result_handle):
+                    if len(blast_record.alignments) > 0:
+                        for alignment in blast_record.alignments:
+                            for hsps in alignment.hsps:
+                                matches = re.split('\+| ', hsps.match)
+                                for match in matches:
+                                    if len(match) >= 8:
```

Comment:
Should this '8' be a parameter (or derived from some other input)?  Though we typically search for 8-11mers, I'm imagining the case where someone only searches for 11mers.  In that case, a length of 11 would be the smallest possible match.  The counterpoint would be that 8-mers are potentially presented by MHC and could theoretically cause problems, but maybe we have 8 as the default and let the user configure it if they want?  Or maybe someone wants to be super-stringent and include anything that matches even a 6-mer. Mostly thinking out loud here -  if it's complex to add, then probably not a huge issue at the moment, but if it's easy, passing it through might be useful?

Reply:
We can make it a parameter. It could either be a new parameter or we could use the lowest epitope length used in the pipeline.